### PR TITLE
Add generator

### PR DESCRIPTION
**Pull request by eyalkraft**
Relates to https://github.com/elastic/security-team/issues/4479

### CODE REVIEW

**Comment by eyalkraft** [comment]
The only part missing is the logic to move the metadata files to their proper places in the bundle.
This is blocked by the policy refactor and thus This PR will be first merged and a follow up will be pushed later on after the policy refactor.

**Comment by eyalkraft** [comment]
Open problems after this PR:
- Some rules metadata end with `    ```\r\n"`
- Some rules have extra backslashes and newlines
```

    aws eks update-cluster-config \\

    \ --region '${REGION_CODE}' \\

    \ --name '${CLUSTER_NAME}' \\

    \ --logging
```
instead of
```
    aws eks update-cluster-config \
     --region '${REGION_CODE}' \
     --name '${CLUSTER_NAME}' \
     --logging
```

- As previously stated, putting the metadata files where they belong in the bundle.

File: cis_policies_generator/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+function normalizeResults(data: BenchmarksData[], benchmark_metadata: BenchmarkMetadata,
+                          profile_applicability: string): Promise<RuleSchema[]> {
+    const sections = parseAllSections(data);
+
+    let result = _.filter(data, (it) => {
+        return Boolean(it["recommendation #"]);
+    });
+
+    return Promise.map(result, (it) => {
```
Comment (uri-weisman) [inline cis_policies_generator/src/index.ts:67]: I'm not quite sure promises are really needed in this use case, we do not perform any async operation.

File: cis_policies_generator/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    for (let idx = 1; idx < tab.data.length; idx++) {
+        const values = tab.data[idx];
+        // `keys` is an array that holds all the column names
+        // `values` is an array that holds the cell values
+        // The following line will push into results an object that is look something like:
+        // {key1: value1, key2: value2, key3: value3, ...}
+        results.push(Object.assign.apply({}, keys.map((v, i) => ({
+            [v]: values[i]?.replace(/\r\n/g, "\n")
+        }))));
+        // results.push(Object.assign.apply({}, keys.map((v, i) => ({[v]: values[i]}))));
```
Comment (uri-weisman) [inline cis_policies_generator/src/index.ts:105]: can be removed?

File: cis_policies_generator/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                    "references": references,
+                    "section": identifySection(it["section #"], sections),
+                    "benchmark": {"name": benchmark_metadata.name, "version": benchmark_metadata.version},
+                };
+                return result;
+            });
+    });
+}
+
+function parseSpreadsheet(tab, benchmark_metadata: BenchmarkMetadata): Promise<RuleSchema[]> {
```
Comment (uri-weisman) [inline cis_policies_generator/src/index.ts:92]: Promise<RuleSchema[]> => RuleSchema[] ?

**Review by uri-weisman — CHANGES_REQUESTED** [review]
Great effort 💪 
Small note, not sure the use of promises is necessary in most functions.

File: cis_policies_generator/src/types.d.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    benchmark: BenchmarkMetadata;
+    default_value?: string;
+    description: string;
+    id: string;
+    impact: string;
+    name: string;
+    profile_applicability: string;
+    rationale: string;
+    references: string[];
+    // rego_rule_id: string;
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/src/types.d.ts:60]: This becomes required in the next task when generating rule templates

File: cis_policies_generator/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
+import _ from "underscore";
+import fs from 'fs';
+import config from 'config';
+import xlsx from 'node-xlsx';
+import path from 'path';
+// import axios from 'axios';
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/src/index.ts:6]: done

File: cis_policies_generator/src/types.d.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    "title": string;
+    "description": string;
+    "rationale statement": string;
+    "impact statement": string;
+    "remediation procedure": string;
+    "audit procedure": string;
+    "references": string;
+}
+
+interface CIS_K8S_Benchmark_Schema {
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/src/types.d.ts:18]: done

File: cis_policies_generator/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+const output_folder: string = config.get("output_folder");
+const benchmarks_folder: string = config.get("benchmarks_folder");
+
+function generateOutputFolder(): void {
+    console.log("Creating output folder:", output_folder);
+    fs.rmdirSync(output_folder, {recursive: true});
+    fs.mkdirSync(output_folder);
+}
+
+function parseReferences(references): Promise<string[]> {
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/src/index.ts:21]: 1. Yes
2. fixing it

File: cis_policies_generator/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
+import _ from "underscore";
+import fs from 'fs';
+import config from 'config';
+import xlsx from 'node-xlsx';
+import path from 'path';
+// import axios from 'axios';
+import Promise from "bluebird";
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/src/index.ts]: fixed

File: cis_policies_generator/src/index.ts (hunk 1 above)
Comment (eyalkraft) [reply cis_policies_generator/src/index.ts]: true

File: cis_policies_generator/package.json [hunk 9]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "cis_policies_generator",
+  "version": "1.0.0",
+  "description": "",
+  "main": "src/index.ts",
+  "scripts": {
+    "start": "set TS_NODE_FILES=true & node -r ts-node/register/transpile-only src/index.ts",
+    "nodemon": "nodemon --watch src/**/*.ts --watch src/**/*.js --exec node -r ts-node/register/transpile-only src/index.ts",
```
Comment (orouz) [inline cis_policies_generator/package.json:8]: `nodemon` should probably be a dev dependency so a single `npm install` would get all deps

File: cis_policies_generator/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+import YAML from 'yaml';
+
+const {v5: uuidv5} = require('uuid');
+
+const output_folder: string = config.get("output_folder");
+const benchmarks_folder: string = config.get("benchmarks_folder");
+
+function generateOutputFolder(): void {
+    console.log("Creating output folder:", output_folder);
+    fs.rmSync(output_folder, {recursive: true});
```
Comment (orouz) [inline cis_policies_generator/src/index.ts:15]:
```suggestion
    if (fs.existsSync(output_folder)) fs.rmSync(output_folder, {recursive: true});
```

File: cis_policies_generator/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
+import _ from "underscore";
+import fs from 'fs';
+import config from 'config';
+import xlsx from 'node-xlsx';
+import path from 'path';
+import YAML from 'yaml';
+
+const {v5: uuidv5} = require('uuid');
```
Comment (orouz) [inline cis_policies_generator/src/index.ts:8]:
```suggestion
import { v5 as uuidv5 } from 'uuid';
```

also:
```npm i --save-dev @types/uuid```

**Review by uri-weisman — APPROVED** [review]

**Review by orouz — COMMENTED** [review]
very cool! 

inlined some comments, mostly on configurations 

we could do more polishing by removing some deps, enabling `noImplicitAny` and a bit of restructuring. not sure it's that critical tho

File: cis_policies_generator/package.json [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  },
+  "dependencies": {
+    "axios": "^0.27.2",
+    "axios-retry": "^3.3.0",
+    "bluebird": "^3.7.2",
+    "config": "^3.3.7",
+    "json-to-pretty-yaml": "^1.2.2",
+    "node-xlsx": "^0.21.0",
+    "underscore": "^1.13.4",
+    "uuid": "^8.3.2",
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/package.json:30]: The uuid code is subject to a change in the near future so I'll leave it that way for now

File: cis_policies_generator/package.json [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    "ts-node": "^10.9.1"
+  },
+  "dependencies": {
+    "axios": "^0.27.2",
+    "axios-retry": "^3.3.0",
+    "bluebird": "^3.7.2",
+    "config": "^3.3.7",
+    "json-to-pretty-yaml": "^1.2.2",
+    "node-xlsx": "^0.21.0",
+    "underscore": "^1.13.4",
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/package.json:29]: I tried getting rid of it, do you have a nice alternative for `_.pick` ?

[the remark this answers is not in the document]
Comment (orouz) [reply cis_policies_generator/package.json:29]:
```diff
- .map(i => _.pick(i, ["section #", "title"]));
+ map(v => ({title: v.title, "section #": v["section #"]}))
```
or 
```diff
+ map(({title, "section #": section}) => ({title, "section #": section}))
```

note it'd be nicer without `section #` (`section_markdown`)

File: cis_policies_generator/tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,32 @@
+{
+  "compilerOptions": { /* Docs: https://www.typescriptlang.org/tsconfig */
+    "target": "ES2020",
+    "module": "commonjs",
+    "outDir": "bin",
+    "rootDir": "src",
+    "strict": true,
+    "noImplicitAny": false,
```
[the remark this answers is not in the document]
Comment (eyalkraft) [reply cis_policies_generator/tsconfig.json:8]: fixed most

**Pull request merged**